README.markdown
# timeframe

A Ruby class for describing and interacting with timeframes.

## Real-world usage

<p><a href="http://brighterplanet.com"><img src="https://s3.amazonaws.com/static.brighterplanet.com/assets/logos/flush-left/inline/green/rasterized/brighter_planet-160-transparent.png" alt="Brighter Planet logo"/></a></p>

We use `timeframe` for [data science at Brighter Planet](http://brighterplanet.com/research) and in production at

* [Brighter Planet's impact estimate web service](http://impact.brighterplanet.com)
* [Brighter Planet's reference data web service](http://data.brighterplanet.com)

Originally proposed to us by [the awesome programmers at fingertips](http:/fngtps.com)

## Based on ISO 8601

As [documented by wikipedia](http://en.wikipedia.org/wiki/ISO_8601#Time_intervals), time intervals are like:

1. Start and end, such as `2007-03-01T13:00:00Z/2008-05-11T15:30:00Z`
2. Start and duration, such as `2007-03-01T13:00:00Z/P1Y2M10DT2H30M`
3. Duration and end, such as `P1Y2M10DT2H30M/2008-05-11T15:30:00Z`
4. Duration only, such as `P1Y2M10DT2H30M`, with additional context information [not supported]

or more simply

    <start>/<end>
    <start>/<duration>
    <duration>/<end>
    <duration> [not supported]

## Precision

Currently the end result is precise to 1 day, so these are the same:

* `2007-03-01T00:00:00Z/2008-05-11T00:00:00Z`
* `2007-03-01/2008-05-11`

This may change in the future.

## Documentation

http://rdoc.info/projects/rossmeissl/timeframe

## Acknowledgements

Thanks to @artemk for https://github.com/rossmeissl/timeframe/pull/5

## Copyright

Copyright (c) 2012 Andy Rossmeissl, Seamus Abshere


CHANGELOG
1.0.0 / 2017-06-22

* Breaking changes

  * No more Array#multiple_timeframes_gaps_left_by (who knows what that was for)

* Enhancements

  * Only uses active_support/core_ext/date
  * Convert to rspec, modernize

0.2.1 / 2012-04-24

* Bug fixes

  * Stop defining Timeframe#to_json... it's conventional to only define #as_json

0.2.0 / 2012-04-24

* Breaking changes

  * Use ISO8601 for JSON output. No sense in coming up with our own structure. Timeframe.parse will still understand the old startDate/endDate hash structure.

* Enhancements

  * New MultiJson syntax
  * Simplify and DRY.
  * Tested on MRI 1.8, MRI 1.9, and JRuby 1.6.7+


test/helper.rb
require 'rubygems'
require 'bundler/setup'

require 'minitest/spec'
require 'minitest/autorun'
require 'minitest/reporters'
MiniTest::Unit.runner = MiniTest::SuiteRunner.new
MiniTest::Unit.runner.reporters << MiniTest::Reporters::SpecReporter.new

require 'timeframe'
require 'timeframe/core_ext/array'


spec/spec_helper.rb
require 'timeframe'

RSpec.configure do |config|
end


spec/timeframe_spec.rb
require_relative 'spec_helper'

describe Timeframe do
  describe 'initialization' do
    it 'should create a timeframe using date strings' do
      tf = Timeframe.new('2008-02-14', '2008-05-10')
      expect(tf.from).to eq(Date.parse('2008-02-14'))
      expect(tf.to).to eq(Date.parse('2008-05-10'))
    end
    it 'should create a timeframe using date objects' do
      start = Date.parse('2008-02-14')
      finish = Date.parse('2008-05-10')
      tf = Timeframe.new(start, finish)
      expect(tf.from).to eq(start)
      expect(tf.to).to eq(finish)
    end
    it "should accept months" do
      timeframe = Timeframe.new(:month => 1)
      expect(timeframe.from).to eq(Date.today.change(:month => 1, :day => 1))
      expect(timeframe.to).to eq(Date.today.change(:month => 2, :day => 1))
    end
    it "should accept month names" do
      timeframe = Timeframe.new(:month => 'february')
      expect(timeframe.from).to eq(Date.today.change(:month => 2, :day => 1))
      expect(timeframe.to).to eq(Date.today.change(:month => 3, :day => 1))
    end
    it "should accept years" do
      timeframe = Timeframe.new(:year => 2004)
      expect(timeframe.from).to eq(Date.new(2004, 1, 1))
      expect(timeframe.to).to eq(Date.new(2005, 1, 1))
    end
    it "should accept years and months" do
      timeframe = Timeframe.new(:year => 2005, :month => 5)
      expect(timeframe.from).to eq(Date.new(2005, 5, 1))
      expect(timeframe.to).to eq(Date.new(2005, 6, 1))
    end
    it "should not accept just one date argument" do
      expect {
        Timeframe.new Date.new(2007, 2, 1)
      }.to raise_error(ArgumentError, /supply/)
    end
    it "should not accept end date that is earlier than start date" do
      expect {
        timeframe = Timeframe.new Date.new(2008, 1, 1), Date.new(2007, 1, 1)
      }.to raise_error(ArgumentError, /earlier/)
    end
    it "should always accept timeframes that cross year boundaries" do
      timeframe = Timeframe.new Date.new(2007, 1, 1), Date.new(2008, 1, 2)
      expect(timeframe.start_date).to eq(Date.new(2007, 1, 1))
    end
  end

  describe '#inspect' do
    it 'should return the time frame in readable text' do
      start = Date.parse('2008-02-14')
      finish = Date.parse('2008-05-10')
      tf = Timeframe.new(start, finish)
      expect(tf.inspect).to match(%r{<Timeframe\(-?\d+\) 86 days starting 2008-02-14 ending 2008-05-10>})
    end
  end

  describe :constrained_new do
    let(:start) { Date.parse('2008-02-14') }
    let(:finish) { Date.parse('2008-05-10') }
    let(:constraint_start) { Date.parse('2008-01-01') }
    let(:constraint_finish) { Date.parse('2008-12-01') }
    let(:constraint) { Timeframe.new(constraint_start, constraint_finish) }

    it "should allow for constrained creation" do
      constraint = Timeframe.new :year => 2008
      may = Timeframe.new Date.new(2008,5,1), Date.new(2008,6,1)
      january = Timeframe.new Date.new(2008,1,1), Date.new(2008,2,1)
      expect(Timeframe.constrained_new(may.from, may.to, constraint)).to eq(may)
      expect(Timeframe.constrained_new(Date.new(2007,1,1), Date.new(2010,1,1), constraint)).to eq(constraint)
      expect(Timeframe.constrained_new(Date.new(2007,11,1), Date.new(2008,2,1), constraint)).to eq(january)
    end
    it 'should return a timeframe spanning start and end date if within constraint' do
      tf = Timeframe.constrained_new(start, finish, constraint)
      expect(tf.from).to eq(start)
      expect(tf.to).to eq(finish)
    end
    it 'should return a timeframe spanning constraint start and end date if outside constraint' do
      constraint = Timeframe.new(start, finish)
      tf = Timeframe.constrained_new(constraint_start, constraint_finish, constraint)
      expect(tf.from).to eq(start)
      expect(tf.to).to eq(finish)
    end
    it 'should return a timeframe starting at constraint start' do
      start = Date.parse('2008-01-01')
      constraint_start = Date.parse('2008-01-14')
      constraint = Timeframe.new(constraint_start, constraint_finish)
      tf = Timeframe.constrained_new(start, finish, constraint)
      expect(tf.from).to eq(constraint_start)
      expect(tf.to).to eq(finish)
    end
    it 'should return a timeframe ending at constraint end' do
      constraint_finish = Date.parse('2008-04-14')
      constraint = Timeframe.new(constraint_start, constraint_finish)
      tf = Timeframe.constrained_new(start, finish, constraint)
      expect(tf.from).to eq(start)
      expect(tf.to).to eq(constraint_finish)
    end
    it "should return a 0-length timeframe when constraining a timeframe by a disjoint timeframe" do
      constraint = Timeframe.new :year => 2010
      timeframe = Timeframe.constrained_new(
        Date.new(2009,1,1), Date.new(2010,1,1), constraint)
      expect(timeframe.days).to eq(0)
    end
  end

  describe :this_year do
    it "should return the current year" do
      expect(Timeframe.this_year).to eq(Timeframe.new(:year => Time.now.year))
    end
  end

  describe '#days' do
    it "should return them number of days included" do
      #TODO: make these separate "it" blocks, per best practices
      expect(Timeframe.new(Date.new(2007, 1, 1), Date.new(2008, 1, 1)).days).to eq(365)
      expect(Timeframe.new(Date.new(2008, 1, 1), Date.new(2009, 1, 1)).days).to eq(366) # leap year
      expect(Timeframe.new(Date.new(2007, 11, 1), Date.new(2007, 12, 1)).days).to eq(30)
      expect(Timeframe.new(Date.new(2007, 11, 1), Date.new(2008, 1, 1)).days).to eq(61)
      expect(Timeframe.new(Date.new(2007, 2, 1), Date.new(2007, 3, 1)).days).to eq(28)
      expect(Timeframe.new(Date.new(2008, 2, 1), Date.new(2008, 3, 1)).days).to eq(29)
      expect(Timeframe.new(Date.new(2008, 2, 1), Date.new(2008, 2, 1)).days).to eq(0)
    end
  end

  describe '#include?' do
    it "should know if a certain date is included in the Timeframe" do
      #TODO: make these separate "it" blocks, per best practices
      timeframe = Timeframe.new :year => 2008, :month => 2
      [
        Date.new(2008, 2, 1),
        Date.new(2008, 2, 5),
        Date.new(2008, 2, 29)
      ].each do |date|
        expect(timeframe.include?(date)).to eq(true)
      end
      [
        Date.new(2008, 1, 1),
        Date.new(2007, 2, 1),
        Date.new(2008, 3, 1)
      ].each do |date|
        expect(timeframe.include?(date)).to eq(false)
      end
    end
  end

  describe '#==' do
    it "should be able to know if it's equal to another Timeframe object" do
      expect(Timeframe.new(:year => 2007)).to eq(Timeframe.new(:year => 2007))
      expect(Timeframe.new(:year => 2004, :month => 1)).to eq(Timeframe.new(:year => 2004, :month => 1))
    end

    it "should hash equal hash values when the timeframe is equal" do
      expect(Timeframe.new(:year => 2007).hash).to eq(Timeframe.new(:year => 2007).hash)
      expect(Timeframe.new(:year => 2004, :month => 1).hash).to eq(Timeframe.new(:year => 2004, :month => 1).hash)
    end
  end

  describe '#months' do
    it "should return an array of month-long subtimeframes" do
      expect(Timeframe.new(:year => 2009).months.length).to eq(12)
    end
  end

  describe '#year' do
    it "should return the relevant year of a timeframe" do
      expect(Timeframe.new(Date.new(2009, 2, 1), Date.new(2009, 4, 1)).year).to eq(Timeframe.new(:year => 2009))
    end
    it "should not return the relevant year of a timeframe if provided an inappropriate range" do
      expect {
        Timeframe.new(Date.new(2009, 1, 1), Date.new(2012, 1, 1)).year
      }.to raise_error(ArgumentError, /dangerous during/)
    end
  end

  describe '#&' do
    it "should return its intersection with another timeframe" do
      #TODO: make these separate "it" blocks, per best practices
      expect(Timeframe.new(:month => 4) & Timeframe.new(:month => 6)).to be_nil
      expect(Timeframe.new(:month => 4) & Timeframe.new(:month => 5)).to be_nil
      expect((Timeframe.new(:month => 4) & Timeframe.new(:month => 4))).to eq(Timeframe.new(:month => 4))
      expect((Timeframe.new(:year => Time.now.year) & Timeframe.new(:month => 4))).to eq(Timeframe.new(:month => 4))
      expect((Timeframe.new(Date.new(2009, 2, 1), Date.new(2009, 6, 1)) & Timeframe.new(Date.new(2009, 4, 1), Date.new(2009, 8, 1)))).to eq(Timeframe.new(Date.new(2009, 4, 1), Date.new(2009, 6, 1)))
    end
  end

  describe '#/' do
    it "should return a fraction of another timeframe" do 
      expect(Timeframe.new(:month => 4, :year => 2009) / Timeframe.new(:year => 2009)).to eq(30.0 / 365.0)
    end
  end

  describe '#gaps_left_by' do
    it "should be able to ascertain gaps left by a list of other Timeframes" do
      expect(Timeframe.new(:year => 2009).gaps_left_by(
        Timeframe.new(:year => 2009, :month => 3),
        Timeframe.new(:year => 2009, :month => 5),
        Timeframe.new(Date.new(2009, 8, 1), Date.new(2009, 11, 1)),
        Timeframe.new(Date.new(2009, 9, 1), Date.new(2009, 10, 1))
      )).to eq(
        [ Timeframe.new(Date.new(2009, 1, 1), Date.new(2009, 3, 1)),
          Timeframe.new(Date.new(2009, 4, 1), Date.new(2009, 5, 1)),
          Timeframe.new(Date.new(2009, 6, 1), Date.new(2009, 8, 1)),
          Timeframe.new(Date.new(2009, 11, 1), Date.new(2010, 1, 1)) ])
    end
  end

  describe '#covered_by?' do
    it "should be able to ascertain gaps left by a list of other Timeframes" do
      expect(Timeframe.new(:year => 2009).covered_by?(
        Timeframe.new(:month => 1, :year => 2009),
        Timeframe.new(Date.new(2009, 2, 1), Date.new(2010, 1, 1))
      )).to eq(true)
    end
  end

  describe '#last_year' do
    it "should return its predecessor in a previous year" do
      expect(Timeframe.this_year.last_year).to eq(Timeframe.new(Date.new(Date.today.year - 1, 1, 1), Date.new(Date.today.year, 1, 1)))
    end
  end
  
  describe :parse do
    describe 'ISO 8601 <start>/<end>' do
      it 'works without time' do
        expect(Timeframe.parse('2007-03-01/2008-05-11')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
        expect(Timeframe.parse('2007-03-01--2008-05-11')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
      end
      it 'works with time' do
        expect(Timeframe.parse('2007-03-01T13:00:00Z/2008-05-11T15:30:00Z')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
        expect(Timeframe.parse('2007-03-01T13:00:00Z--2008-05-11T15:30:00Z')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
      end
      it 'takes shorthand' do
        expect(Timeframe.parse('2007-11-13/15')).to eq(Timeframe.new(Date.new(2007, 11, 13), Date.new(2007, 11, 15))) # "2007-11-13/15", i.e. from any time on 2007-11-13 to any time on 2007-11-15
        expect(Timeframe.parse("2008-02-15/03-14")).to eq(Timeframe.new(Date.new(2008, 2, 15), Date.new(2008, 3, 14))) # "2008-02-15/2008-03-14"
        expect(Timeframe.parse("2007-12-14T13:30/15:30")).to eq(Timeframe.new(Date.new(2007, 12, 14), Date.new(2007, 12, 14))) # "2007-12-14T13:30/2007-12-14T15:30".. imprecise!
      end
    end

    describe 'ISO 8601 <start>/<duration>' do
      it 'works without time' do
        expect(Timeframe.parse('2007-03-01/P1Y2M10DT2H30M')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
        expect(Timeframe.parse('2007-03-01--P1Y2M10DT2H30M')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
      end
      it 'works with time' do
        expect(Timeframe.parse('2007-03-01T13:00:00Z/P1Y2M10DT2H30M')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
        expect(Timeframe.parse('2007-03-01T13:00:00Z--P1Y2M10DT2H30M')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 5, 11)))
      end
    end
    
    # note that 2008 was a leap year
    describe 'ISO 8601 <duration>/<end>' do
      it 'works with leap years' do
        expect(Timeframe.parse('2007-02-28--P1Y')).to eq(Timeframe.new(Date.new(2007, 2, 28), Date.new(2008, 2, 29)))
        expect(Timeframe.parse('P1Y--2008-02-29')).to eq(Timeframe.new(Date.new(2007, 3, 1), Date.new(2008, 2, 29)))
      end
      it 'works without time' do
        expect(Timeframe.parse('P1Y2M10DT2H30M/2008-05-11')).to eq(Timeframe.new(Date.new(2007, 3, 2), Date.new(2008, 5, 11)))
        expect(Timeframe.parse('P1Y2M10DT2H30M--2008-05-11')).to eq(Timeframe.new(Date.new(2007, 3, 2), Date.new(2008, 5, 11)))
      end
      it 'works with time' do
        expect(Timeframe.parse('P1Y2M10DT2H30M/2008-05-11T15:30:00Z')).to eq(Timeframe.new(Date.new(2007, 3, 3), Date.new(2008, 5, 11)))
        expect(Timeframe.parse('P1Y2M10DT2H30M--2008-05-11T15:30:00Z')).to eq(Timeframe.new(Date.new(2007, 3, 3), Date.new(2008, 5, 11)))
      end
    end

    it 'understands plain year' do
      plain_year = 2009
      expect(Timeframe.parse(plain_year)).to eq(Timeframe.new(:year => plain_year))
      expect(Timeframe.parse(plain_year.to_s)).to eq(Timeframe.new(:year => plain_year))
    end
    it 'understands JSON' do
      json =<<-EOS
      "2009-05-01/2009-06-01"
EOS
      expect(Timeframe.parse(json)).to eq(Timeframe.new(:year => 2009, :month => 5))
    end
    it 'understands a particular style of Ruby hash we used to emit (deprecated)' do
      hsh = { :startDate => '2009-05-01', :endDate => '2009-06-01' }
      expect(Timeframe.parse(hsh)).to eq(Timeframe.new(:year => 2009, :month => 5))
      expect(Timeframe.parse(hsh.stringify_keys)).to eq(Timeframe.new(:year => 2009, :month => 5))
    end
  end

  describe '#as_json' do
    it 'should generate JSON (test fails on ruby 1.8)' do
      expect(Timeframe.new(:year => 2009).as_json).to eq(%{2009-01-01/2010-01-01})
      expect(MultiJson.dump(Timeframe.new(:year => 2009))).to eq(Timeframe.new(:year => 2009).as_json.inspect)
    end
    it 'understands its own JSON' do
      t = Timeframe.new(:year => 2009, :month => 5)
      expect(Timeframe.from_json(MultiJson.dump(t))).to eq(t)
    end
  end

  describe '#to_param' do
    it 'should generate a URL-friendly parameter' do
      expect(Timeframe.new(:year => 2009).to_param).to eq("2009-01-01/2010-01-01")
    end
    it 'understands its own #to_param' do
      t = Timeframe.new(:year => 2009, :month => 5)
      expect(Timeframe.parse(t.to_param)).to eq(t)
    end
  end

  describe '#to_s' do
    it 'should not only look at month numbers when describing multi-year timeframes' do
      expect(Timeframe.new(Date.parse('2008-01-01'), Date.parse('2010-01-01')).to_s).to eq("2008-01-01/2010-01-01")
    end
  end
  
  describe '#dates' do
    it "should enumerate all dates between start and end" do
      dates = Timeframe.new(:year => 2008).dates
      expect(dates.min).to eq(Date.new(2008,1,1))
      expect(dates.max).to eq(Date.new(2008,12,31))
      expect(dates.uniq.length).to eq(366)
      expect(dates.select { |d| d.month == 2 }.length).to eq(29)
    end
  end
  
  describe '#first_days_of_months' do
    it "should enumerate all the first days of included months" do
      dates = Timeframe.parse('2011-05-01/2012-02-01').first_days_of_months
      expect(dates).to eq([Date.new(2011,5,1), Date.new(2011,6,1), Date.new(2011,7,1), Date.new(2011,8,1), Date.new(2011,9,1), Date.new(2011,10,1), Date.new(2011,11,1), Date.new(2011,12,1), Date.new(2012,1,1)])
    end
  end
end


lib/timeframe/iso_8601.rb
class Timeframe
  module Iso8601
    # Internal use.
    #
    # Parses a duration like 'P1Y2M4DT3H4M2S'
    class Duration
      attr_reader :seconds
      def initialize(date_part, time_part)
        y       = parse date_part, :Y
        m       = parse date_part, :M
        d       = parse date_part, :D
        h       = parse time_part, :H
        minutes = parse time_part, :M
        s       = parse time_part, :S
        @seconds = (y*31_556_926 + m*2_629_743.83 + d*86_400 + h*3_600 + minutes*60 + s).ceil
      end
      def parse(part, indicator)
        if part =~ /(\d+)#{indicator.to_s}/
          $1.to_f
        else
          0
        end
      end
    end

    # Internal use.
    class Side
      # We add one day because so that it can be excluded per timeframe's conventions.
      EXCLUDED_LAST_DAY = 86_400
      attr_reader :date_part
      attr_reader :time_part
      def to_time(counterpart)
        if date_part.start_with?('P')
          counterpart.resolve_time(self) + offset + EXCLUDED_LAST_DAY
        else
          resolve_time counterpart
        end
      end
    end
    
    # Internal use.
    #
    # The "A" side of "A/B"
    class A < Side
      def initialize(raw)
        raw = raw.upcase
        @date_part, @time_part = raw.split('T')
        @time_part ||= ''
      end
      def resolve_time(*)
        Time.parse [date_part, time_part].join('T')
      end
      # When A is a period, it counts as a negative offset to B.
      def offset
        0.0 - Duration.new(date_part, time_part).seconds
      end
    end
    
    # Internal use.
    #
    # The "B" side of "A/B"
    class B < Side
      def initialize(raw)
        raw = raw.upcase
        if raw.include?(':') and not raw.include?('T')
          # it's just a shorthand for time
          @date_part = ''
          @time_part = raw
        else
          @date_part, @time_part = raw.split('T')
          @time_part ||= ''
        end
      end
      # When shorthand is used, we need to peek at our counterpart (A) in order to steal letters
      # Shorthand can only be used on the B side, and only in <start>/<end> format.
      def resolve_time(counterpart)
        filled_in_date_part = unless date_part.count('-') == 2
          counterpart.date_part[0..(0-date_part.length-1)] + date_part
        else
          date_part
        end
        filled_in_time_part = if time_part.count(':') < 2
          counterpart.time_part[0..(0-time_part.length-1)] + time_part
        else
          time_part
        end
        Time.parse [filled_in_date_part, filled_in_time_part].join('T')
      end
      def offset
        Duration.new(date_part, time_part).seconds
      end
    end
  end
end


lib/timeframe.rb
require 'date'
require 'multi_json'
require 'active_support'
require 'active_support/core_ext/date'

require 'timeframe/iso_8601'

# Encapsulates a timeframe between two dates. The dates provided to the class are always until the last date. That means
# that the last date is excluded.
#
#   # from 2007-10-01 00:00:00.000 to 2007-10-31 23:59:59.999
#   Timeframe.new(Date(2007,10,1), Date(2007,11,1))
#   # and holds 31 days
#   Timeframe.new(Date(2007,10,1), Date(2007,11,1)).days #=> 31
class Timeframe
  class << self    
    # Shortcut method to return the Timeframe representing the current year (as defined by Time.now)
    def this_year
      new :year => Time.now.year
    end
    
    # Construct a new Timeframe, but constrain it by another
    def constrained_new(start_date, end_date, constraint)
      start_date, end_date = make_dates start_date, end_date
      raise ArgumentError, 'Constraint must be a Timeframe' unless constraint.is_a? Timeframe
      raise ArgumentError, "Start date #{start_date} should be earlier than end date #{end_date}" if start_date > end_date
      if end_date <= constraint.start_date or start_date >= constraint.end_date
        new constraint.start_date, constraint.start_date
      elsif start_date.year == end_date.yesterday.year
        new(start_date, end_date) & constraint
      elsif start_date.year < constraint.start_date.year and constraint.start_date.year < end_date.yesterday.year
        constraint
      else
        new [constraint.start_date, start_date].max, [constraint.end_date, end_date].min
      end
    end
    
    # Create a timeframe +/- number of years around today
    def mid(number)
      start_date = Time.now.today - number.years
      end_date = Time.now.today + number.years
      new start_date, end_date
    end
    
    # Construct a new Timeframe by parsing an ISO 8601 time interval string
    # http://en.wikipedia.org/wiki/ISO_8601#Time_intervals
    def from_iso8601(str)
      delimiter = str.include?('/') ? '/' : '--'
      a_raw, b_raw = str.split delimiter
      if a_raw.blank? or b_raw.blank?
        raise ArgumentError, "Interval must be specified according to ISO 8601 <start>/<end>, <start>/<duration>, or <duration>/<end>."
      end
      a = Iso8601::A.new a_raw
      b = Iso8601::B.new b_raw
      new a.to_time(b), b.to_time(a)
    end
    
    # Construct a new Timeframe from a hash with keys startDate and endDate    
    def from_hash(hsh)
      hsh = hsh.symbolize_keys
      new hsh[:startDate], hsh[:endDate]
    end
    
    # Construct a new Timeframe from a year.
    def from_year(year)
      new :year => year.to_i
    end
    
    # Automagically parse a Timeframe from either a String or a Hash
    def parse(input)
      case input
      when ::Integer
        from_year input
      when ::Hash
        from_hash input
      when ::String
        str = input.strip
        if str.start_with?('{')
          from_hash MultiJson.load(str)
        elsif input =~ /\A\d\d\d\d\z/
          from_year input
        else
          from_iso8601 str
        end
      else
        raise ArgumentError, "Must be String or Hash"
      end
    end
    alias :interval :parse
    alias :from_json :parse
    
    # Deprecated
    def multiyear(*args) # :nodoc:
      new *args
    end

    def to_date(v)
      case v
      when NilClass
        nil
      when Date
        v
      when Time
        v.to_date
      else
        Date.parse v
      end
    end

    private
    
    def make_dates(start_date, end_date)
      [to_date(start_date), to_date(end_date)]
    end
  end

  attr_reader :start_date
  attr_reader :end_date

  # Creates a new instance of Timeframe. You can either pass a start and end Date or a Hash with named arguments,
  # with the following options:
  #
  #   <tt>:month</tt>: Start date becomes the first day of this month, and the end date becomes the first day of
  #   the next month. If no <tt>:year</tt> is specified, the current year is used.
  #   <tt>:year</tt>: Start date becomes the first day of this year, and the end date becomes the first day of the
  #   next year.
  #
  # Examples:
  #
  #   Timeframe.new Date.new(2007, 2, 1), Date.new(2007, 4, 1) # February and March
  #   Timeframe.new :year => 2004 # The year 2004
  #   Timeframe.new :month => 4 # April
  #   Timeframe.new :year => 2004, :month => 2 # Feburary 2004
  def initialize(*args)
    options = args.extract_options!

    if month = options[:month]
      month = Date.parse(month).month if month.is_a? String
      year = options[:year] || Date.today.year
      start_date = Date.new(year, month, 1)
      end_date   = start_date.next_month
    elsif year = options[:year]
      start_date = Date.new(year, 1, 1)
      end_date   = Date.new(year+1, 1, 1)
    end

    start_date ||= Timeframe.to_date(args[0])
    end_date ||= Timeframe.to_date(args[1])

    raise ArgumentError, "Please supply a start and end date, `#{args.map(&:inspect).to_sentence}' is not enough" if start_date.nil? or end_date.nil?
    raise ArgumentError, "Start date #{start_date} should be earlier than end date #{end_date}" if start_date > end_date

    @start_date, @end_date = start_date, end_date
  end

  def inspect # :nodoc:
    "<Timeframe(#{object_id}) #{days} days starting #{start_date} ending #{end_date}>"
  end
    
  # The number of days in the timeframe
  #
  #   Timeframe.new(Date.new(2007, 11, 1), Date.new(2007, 12, 1)).days #=> 30
  #   Timeframe.new(:month => 1).days #=> 31
  #   Timeframe.new(:year => 2004).days #=> 366
  def days
    (end_date - start_date).to_i
  end

  # Returns true when a Date or other Timeframe is included in this Timeframe
  def include?(obj)
    # puts "checking to see if #{date} is between #{start_date} and #{end_date}" if Emitter::DEBUG
    case obj
    when Date
      (start_date...end_date).include?(obj)
    when Time
      # (start_date...end_date).include?(Date.parse(obj))
      raise "this wasn't previously supported, but it could be"
    when Timeframe
      start_date <= obj.start_date and end_date >= obj.end_date
    end
  end

  # Returns true when the parameter Timeframe is properly included in the Timeframe
  def proper_include?(other_timeframe)
    raise ArgumentError, 'Proper inclusion only makes sense when testing other Timeframes' unless other_timeframe.is_a? Timeframe
    (start_date < other_timeframe.start_date) and (end_date > other_timeframe.end_date)
  end

  # Returns true when this timeframe is equal to the other timeframe
  def ==(other)
    # puts "checking to see if #{self} is equal to #{other}" if Emitter::DEBUG
    return false unless other.is_a?(Timeframe)
    start_date == other.start_date and end_date == other.end_date
  end
  alias :eql? :==

  # Calculates a hash value for the Timeframe, used for equality checking and Hash lookups.
  def hash
    start_date.hash + end_date.hash
  end

  # Returns the relevant year as a Timeframe
  def year
    raise ArgumentError, 'Timeframes that cross year boundaries are dangerous during Timeframe#year' unless start_date.year == end_date.yesterday.year
    Timeframe.new :year => start_date.year
  end

  # Returns an Array of month-long Timeframes. Partial months are **not** included by default.
  # http://stackoverflow.com/questions/1724639/iterate-every-month-with-date-objects
  def months
    memo = []
    ptr = start_date
    while ptr <= end_date do
      memo.push(Timeframe.new(:year => ptr.year, :month => ptr.month) & self)
      ptr = ptr >> 1
    end
    memo.flatten.compact
  end

  # Crop a Timeframe to end no later than the provided date.
  def ending_no_later_than(date)
    if end_date < date
      self
    elsif start_date >= date
      nil
    else
      Timeframe.new start_date, date
    end
  end

  # Returns a timeframe representing the intersection of the given timeframes
  def &(other_timeframe)
    this_timeframe = self
    if other_timeframe == this_timeframe
      this_timeframe
    elsif this_timeframe.start_date > other_timeframe.start_date and this_timeframe.end_date < other_timeframe.end_date
      this_timeframe
    elsif other_timeframe.start_date > this_timeframe.start_date and other_timeframe.end_date < this_timeframe.end_date
      other_timeframe
    elsif this_timeframe.start_date >= other_timeframe.end_date or this_timeframe.end_date <= other_timeframe.start_date
      nil
    else
      Timeframe.new [this_timeframe.start_date, other_timeframe.start_date].max, [this_timeframe.end_date, other_timeframe.end_date].min
    end
  end

  # Returns the fraction (as a Float) of another Timeframe that this Timeframe represents
  def /(other_timeframe)
    raise ArgumentError, 'You can only divide a Timeframe by another Timeframe' unless other_timeframe.is_a? Timeframe
    self.days.to_f / other_timeframe.days.to_f
  end

  # Crop a Timeframe by another Timeframe
  def crop(container)
    raise ArgumentError, 'You can only crop a timeframe by another timeframe' unless container.is_a? Timeframe
    self.class.new [start_date, container.start_date].max, [end_date, container.end_date].min
  end

  # Returns an array of Timeframes representing the gaps left in the Timeframe after removing all given Timeframes
  def gaps_left_by(*timeframes)
    # remove extraneous timeframes
    timeframes.reject! { |t| t.end_date <= start_date }
    timeframes.reject! { |t| t.start_date >= end_date }
    
    # crop timeframes
    timeframes.map! { |t| t.crop self }

    # remove proper subtimeframes
    timeframes.reject! { |t| timeframes.detect { |u| u.proper_include? t } }

    # escape
    return [self] if  timeframes.empty?

    timeframes.sort! { |x, y| x.start_date <=> y.start_date }
    
    a = [ start_date ] + timeframes.collect(&:end_date)
    b = timeframes.collect(&:start_date) + [ end_date ]

    a.zip(b).map do |gap|
      Timeframe.new(*gap) if gap[1] > gap[0]
    end.compact
  end

  # Returns true if the union of the given Timeframes includes the Timeframe
  def covered_by?(*timeframes)
    gaps_left_by(*timeframes).empty?
  end

  # Returns the same Timeframe, only a year earlier
  def last_year
    self.class.new(
      Date.new(start_date.year - 1, start_date.month, start_date.day),
      Date.new(end_date.year - 1, end_date.month, end_date.day)
    )
  end
  
  def as_json(*)
    iso8601
  end
    
  # An ISO 8601 "time interval" like YYYY-MM-DD/YYYY-MM-DD
  def iso8601
    "#{start_date.iso8601}/#{end_date.iso8601}"
  end
  alias :to_s :iso8601
  alias :to_param :iso8601
  
  def dates
    dates = []
    cursor = start_date
    while cursor < end_date
      dates << cursor
      cursor = cursor.succ
    end
    dates
  end
  
  def first_days_of_months
    dates = []
    cursor = start_date.beginning_of_month
    while cursor < end_date
      dates << cursor
      cursor = cursor >> 1
    end
    dates
  end
  
  # Deprecated
  def from # :nodoc:
    @start_date
  end
  
  # Deprecated
  def to # :nodoc:
    @end_date
  end
end
